### PR TITLE
added tests for all widgets

### PR DESCRIPTION
**Pull request by superzazu**
Hello!

I added tests for all the widgets, using the test base you implemented. That allowed me to see that two methods were missing (Control.getVisible() and Checkbox.onToggled()), and also allowed me to fix two little bugs (one `return` missing from Control.getEnabled() and a wrong call to `uiSpinboxValue ` in Spinbox.getValue()). I also removed `return` keywords in methods (in control wrappers) that don't return a value.

Best regards :-)

### CODE REVIEW

File: tests/test_button.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        self.button = Button('My button')
+
+    def test_text_attribute(self):
+        """Tests the button text."""
+        text = 'My new button'
+        self.button.setText(text)
+        self.assertEqual(self.button.getText(), text)
+
+    def test_onClick_is_called_when_user_presses_button(self):
+        raise  # TODO: how do we simulate mouse press ?
```
Comment (joaoventura) [inline tests/test_button.py]: Sometimes we may need to use mock (https://docs.python.org/3/library/unittest.mock.html) to try to simulate some situations. However, in this particular case I don't know how we can use it to really press the button.

File: tests/test_checkbox.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        """Tests the checkbox's `text` attribute."""
+        text = 'My new checkbox'
+        self.checkbox.setText(text)
+        self.assertEqual(self.checkbox.getText(), text)
+
+    def test_checked_attribute(self):
+        """Tests the checkbox's `checked` attribute."""
+        self.assertEqual(self.checkbox.getChecked(), False)
+        self.checkbox.setChecked(True)
+        self.assertEqual(self.checkbox.getChecked(), True)
```
Comment (joaoventura) [inline tests/test_checkbox.py]: Ideally, we should only have one assert per test. This means we are testing only one specific thing, and if it fails, we know what really failed.

In this case you should create two tests to validate that we can trigger it to True and then another to validate we can trigger it to false.

File: tests/test_control.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    def setUp(self):
+        super().setUp()
+        self.control = Control(libui.uiNewButton(''))
+
+    def test_visible_attribute(self):
+        """Tests the control's `visible` attribute."""
+        self.assertEqual(self.control.getVisible(), True)
+        self.control.hide()
+        self.assertEqual(self.control.getVisible(), False)
```
Comment (joaoventura) [inline tests/test_control.py]: Same as above. One test condition per test method..

File: tests/test_slider.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+class SliderTest(WindowTestCase):
+
+    def setUp(self):
+        super().setUp()
+        self.slider = Slider(0, 100)
+
+    def test_value_attribute(self):
+        """Tests the slider's `value` attribute."""
+        self.assertEqual(self.slider.getValue(), 0)
```
Comment (joaoventura) [inline tests/test_slider.py]: This assert is for a specific test like "test_initial_value"..

File: tests/test_spinbox.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+class SpinboxTest(WindowTestCase):
+
+    def setUp(self):
+        super().setUp()
+        self.spinbox = Spinbox(0, 100)
+
+    def test_value_attribute(self):
+        """Tests the spinbox's `value` attribute."""
+        self.assertEqual(self.spinbox.getValue(), 0)
```
Comment (joaoventura) [inline tests/test_spinbox.py]: Same as for slider, initial value of spinbox should be zero is a possible test..

File: tests/test_tab.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class TabTest(WindowTestCase):
+
+    def setUp(self):
+        super().setUp()
+        self.tab = Tab()
+
+    def test_can_append_pages_to_tab(self):
+        """Tests that we can append pages to tab."""
```
Comment (joaoventura) [inline tests/test_tab.py]: You should try to improve a little bit the description of the tests. In this case you are not testing that you can append pages to tab, but that the control correctly returns the number of pages. The assert is what you are really testing...

File: tests/test_tab.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+        button2 = Button('second button')
+        self.tab.insertAt('my tab 1', 0, button2)
+
+        # TODO: should we do a check on indexes when user calls `insertAt` ?
+        #       At the moment, the following code crashes:
+        # button3 = Button('second button')
+        # self.tab.insertAt('my tab 3', 10, button3) # non-existing page index
+
+    def test_can_delete_pages_from_tab(self):
```
Comment (joaoventura) [inline tests/test_tab.py]: Also, try to reduce the name of the methods. The description of the method will give more details of the test..

File: tests/test_window.py [hunk 8]
```diff
@@ -9,12 +9,12 @@
 class WindowTest(WindowTestCase):
 
     def test_window_title(self):
-        """Tests the window title"""
+        """Tests the window's title"""
```
Comment (joaoventura) [inline tests/test_window.py]: No need to use correct english with apostrophes here..

**Comment by joaoventura** [comment]
Great work, and good to know that you caught some bugs already with it.. I've added some notes to your pull request, if you can work on them it would be great! 👍

File: tests/test_tab.py (hunk 6 above)
Comment (superzazu) [inline tests/test_tab.py]: Actually, I really wanted to test that I can append pages to tab; however the only way to check the feature was using `getNumPages`.
I agree it is a bit misleading, I'll try to improve this :-)

File: tests/test_tab.py (hunk 7 above)
Comment (superzazu) [inline tests/test_tab.py]: Ok ! How should I call this one ? `test_delete_pages` ?

File: tests/test_window.py (hunk 8 above)
Comment (superzazu) [inline tests/test_window.py]: Sorry, I never know which to use...

File: tests/test_tab.py (hunk 6 above)
Comment (joaoventura) [inline tests/test_tab.py]: Ah ok.. Yes, sometimes the result of one test depends on other factors and that's ok! 
A good and short name for this test could be "test_append_pages" and a description could be something like "Pages can be appended to a tab."

File: tests/test_tab.py (hunk 7 above)
Comment (joaoventura) [inline tests/test_tab.py]: It's a good name, yes! 👍

File: tests/test_window.py (hunk 8 above)
Comment (joaoventura) [inline tests/test_window.py]: Yours is better than mine (i.e., the apostrophe is correct). But in this kind of cases we do not need to follow religiously the english vocabulary, although there's no harm if we do it! 👍

**Comment by superzazu** [comment]
Okay, I tried to fix all these problems. Feel free to comment if anything is not optimal. :-)

**Comment by joaoventura** [comment]
👍

**Pull request merged**